room/compiler-processing/src/main/java/androidx/room/compiler/processing/ksp/KspType.kt

```diff
@@ -0,0 +1,174 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.room.compiler.processing.ksp
+
+import androidx.room.compiler.processing.XDeclaredType
+import androidx.room.compiler.processing.XEquality
+import androidx.room.compiler.processing.XNullability
+import androidx.room.compiler.processing.XType
+import androidx.room.compiler.processing.XTypeElement
+import androidx.room.compiler.processing.rawTypeName
+import com.squareup.javapoet.ClassName
+import com.squareup.javapoet.TypeName
+import org.jetbrains.kotlin.ksp.symbol.KSClassDeclaration
+import org.jetbrains.kotlin.ksp.symbol.KSType
+import org.jetbrains.kotlin.ksp.symbol.KSTypeReference
+import org.jetbrains.kotlin.ksp.symbol.Nullability
+import kotlin.reflect.KClass
+
+/**
+ * XType implementation for KSP type.
+ *
+ * It might be initialized with a [KSTypeReference] or [KSType] depending on the call point.
+ *
+ * We don't necessarily have a [KSTypeReference] (e.g. if we are getting it from an element).
+ * Similarly, we may not be able to get a [KSType] (e.g. if it resolves to error).
+ */
+internal class KspType private constructor(
+    private val env: KspProcessingEnv,
+    private val ksTypeReference: KSTypeReference?,
+    resolved: KSType?
+) : XDeclaredType, XEquality {
+    constructor(env: KspProcessingEnv, ksTypeReference: KSTypeReference) : this(
+        env = env,
+        ksTypeReference = ksTypeReference,
+        resolved = null
+    )
+
+    constructor(env: KspProcessingEnv, resolved: KSType) : this(
+        env = env,
+        ksTypeReference = null,
+        resolved = resolved
+    )
+
+    internal val ksType by lazy {
+        resolved ?: ksTypeReference?.resolve()
+    }
+    override val rawType by lazy {
+        KspRawType(this)
+    }
+    override val typeArguments: List<XType> by lazy {
+        // prioritize type reference if it exists as it will have actaul types user picked.
+        val args = ksTypeReference?.element?.typeArguments
+            ?: resolved?.arguments
+            ?: emptyList()
+        args.map {
+            // TODO what if it.type is null?
+            env.wrap(it.type!!)
+        }
+    }
+    override val typeName: TypeName by lazy {
+        ksType?.typeName() ?: ksTypeReference?.typeName() ?: UNDEFINED
+    }
+    override val nullability by lazy {
+        when (ksType?.nullability) {
+            Nullability.NULLABLE -> XNullability.NULLABLE
+            Nullability.NOT_NULL -> XNullability.NONNULL
+            else -> XNullability.UNKNOWN
```

Comment:
The third nullability we have is PLATFORM nullability, just want to make sure it is what you mean by UNKNOWN.

Reply:
yea platform is what we call UNKNOWN.